<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <!-- Convert an OctoMap published on the given topic to markers and/or point clouds
       with optional ceiling removal.  Particularly useful for bag files. -->

  <!-- Resolution of the input map -->
  <arg name="merged_resolution" default="0.15" />
  <!-- Fixed map frame (set to 'map' if SLAM or localization running!) -->
  <arg name="frame_id" default="world" />
  <!-- Input type: 0=none, 1=pointcloud, 2=octomap, 3=diffs -->
  <arg name="input" default="3" />
  <!-- <arg name="input_topic" default="/H01/octomap_binary" /> -->
  <arg name="input_topic" default="/H01/merged_map" />
  <!-- <arg name="input_topic" default="/Base/neighbors/H01/map" /> -->
  <arg name="neighbors_input_topic" default="/H01/neighbor_maps" />
  <!-- <arg name="neighbors_input_topic" default="/Base/neighbor_maps" /> -->
  <!-- Data source(s) to integrate (OctoMap) -->
  <arg name="agents" default="H01, H02, D01, D02" />
  <arg name="diff_pre" default="/Base/neighbors/" />
  <arg name="diff_post" default="/map" />
  <arg name="map_merge_timer" default="200" />

  <!-- Topic prefix to output to, ie merged_map outputs to merged_map_markers, etc -->
  <arg name="prefix" default="merged_map" />
  <arg name="publish_merged_binary" default="true" />
  <arg name="publish_neighbor_maps" default="true" />
  <!-- Which color mode to use for map publishing.  Options are 1 (agent), 2 (rough), 3 (z-axis) -->
  <arg name="display_color" default="1" />
  <!-- How often to publish these maps (marker array, point cloud) -->
  <arg name="publish_optional_duration" default="1" />
  <!-- Whether to publish the marker arrays and point cloud -->
  <arg name="publish_marker_array" default="false" />
  <arg name="publish_free_space" default="false" />
  <arg name="publish_point_cloud" default="false" />
  <!-- Whether to remove ceiling from marker arrays and point clouds, and how deep to check -->
  <arg name="remove_ceiling" default="true" />
  <arg name="remove_ceiling_depth" default="4" />

  <arg name="enable_traversability"           default="true" />
  <arg name="enable_traversability_sharing"   default="false" />

  <!-- <arg name="color_pairs" default="2:H01,X1;3:H02,X2;4:H03,X3;5:D01,X4,A01;6:D02,X5,A02;7:A03;8:A04;9:A05" /> -->
  <arg name="color_pairs" default="2:H01,X1;4:H02,X2;4:H03,X3;5:D01,X4,A01;6:D02,X5,A02;7:A03;8:A04;9:A05" />

  <node pkg="marble_mapping" type="marble_mapping_node" name="merge_diffs" output="screen">
    <param name="merged_resolution" value="$(arg merged_resolution)" />
    <param name="frame_id" type="string" value="$(arg frame_id)" />
    <param name="input" value="$(arg input)" />
    <param name="input_topic" value="$(arg input_topic)" />
    <param name="neighbors_input_topic" value="$(arg neighbors_input_topic)" />
    <param name="agents" value="$(arg agents)" />
    <param name="diff_pre" value="$(arg diff_pre)" />
    <param name="diff_post" value="$(arg diff_post)" />
    <param name="map_merge_timer" value="$(arg map_merge_timer)" />

    <param name="publish_merged_binary" value="$(arg publish_merged_binary)" />
    <param name="publish_neighbor_maps" value="$(arg publish_neighbor_maps)" />
    <param name="publish_optional_duration" value="$(arg publish_optional_duration)" />
    <param name="publish_marker_array" value="$(arg publish_marker_array)" />
    <param name="publish_free_space" value="$(arg publish_free_space)" />
    <param name="publish_point_cloud" value="$(arg publish_point_cloud)" />

    <param name="remove_ceiling" value="$(arg remove_ceiling)" />
    <param name="remove_ceiling_depth" value="$(arg remove_ceiling_depth)" />

    <remap from="merged_map_markers" to="$(arg prefix)_markers" />
    <remap from="merged_map_free" to="$(arg prefix)_free" />
    <remap from="merged_map_pc" to="$(arg prefix)_pc" />

    <param name="enable_traversability"           value="$(arg enable_traversability)" />
    <param name="enable_traversability_sharing"   value="$(arg enable_traversability_sharing)" />

    <param name="color_pairs" value="$(arg color_pairs)" />
  </node>
  <!-- Open RViz with config file -->
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find octomap_merging)/rviz/base_station_config.rviz"> </node>
</launch>